src/Compilers/CSharp/Test/Syntax/Parsing/RefStructs.cs

```diff
@@ -0,0 +1,106 @@
+// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Xunit;
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Roslyn.Test.Utilities;
+using Xunit.Abstractions;
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests.Parsing
+{
+    [CompilerTrait(CompilerFeature.ReadonlyReferences)]
+    public class RefStructs : ParsingTests
+    {
+        public RefStructs(ITestOutputHelper output) : base(output) { }
+
+        protected override SyntaxTree ParseTree(string text, CSharpParseOptions options)
+        {
+            return SyntaxFactory.ParseSyntaxTree(text, options: options);
+        }
+
+        [Fact]
+        public void RefStructSimple()
+        {
+            var text = @"
+class Program
+{
+    ref struct S1{}
+
+    public ref struct S2{}
+}
+";
+
+            var comp = CreateCompilationWithMscorlib45(text, parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.Latest), options: TestOptions.DebugDll);
+            comp.VerifyDiagnostics(
+            );
+        }
+
+        [Fact]
+        public void RefStructSimpleLangVer()
+        {
+            var text = @"
+class Program
+{
+    ref struct S1{}
+
+    public ref struct S2{}
+}
+";
+
+            var comp = CreateCompilationWithMscorlib45(text, parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.CSharp7), options: TestOptions.DebugDll);
+            comp.VerifyDiagnostics(
+                // (4,5): error CS8107: Feature 'ref structs' is not available in C# 7. Please use language version 7.1 or greater.
+                //     ref struct S1{}
+                Diagnostic(ErrorCode.ERR_FeatureNotAvailableInVersion7, "ref").WithArguments("ref structs", "7.1").WithLocation(4, 5),
+                // (6,12): error CS8107: Feature 'ref structs' is not available in C# 7. Please use language version 7.1 or greater.
+                //     public ref struct S2{}
+                Diagnostic(ErrorCode.ERR_FeatureNotAvailableInVersion7, "ref").WithArguments("ref structs", "7.1").WithLocation(6, 12)
+            );
+        }
+
+        [Fact]
+        public void RefStructErr()
+        {
+            var text = @"
```

Comment:
also added a delegate

---
In reply to: [116117586](https://github.com/dotnet/roslyn/pull/19411#discussion_r116117586) [](ancestors = 116117586)